src/Space.cpp

```diff
@@ -1071,9 +1068,18 @@ void Space::TimeStep(float step)
 		b->UpdateFrame();
 
 	// AI acts here, then move all bodies and frames
-	for (Body *b : m_bodies)
+	// NOTE: The AI can add bodies here so we can't use an iterator
+	// this restores the previous version where only the initial list is
+	// updated unless the bodies vector reallocated where anything could
+	// have happened
+	// alternative fixes to delay addition caused
+	// https://github.com/pioneerspacesim/pioneer/issues/5695
+	//
+	// THIS IS A HACK/WORKAROUND until a more proper solution can be found
+	for (size_t i = 0, e = m_bodies.size(); i != e; ++i) {
```

Comment:
As far as I can tell, this will retain the old behavior where only the bodies present at the start of iteration are actually ticked here. To iterate over new-this-frame bodies, the condition needs to be changed to `i != m_bodies.size()` - `e` is evaluated once at the start of the loop and if new bodies are added to the end they will not be updated.

Reply:
So... the original behavior used an 

`for (auto b : m_bodies )`

Which expands to 

`for (auto i = m_bodies.begin(), auto e = m_bodies.end(), i != e; ++i ) { auto b = *i; `

So I went for the same logic, which means that their static update wouldn't get called as I wanted to make the code behave as close as possible to how it did before, without the risk of having invalid iterators.

After our IRC Chat, I agree, we should check `size()` each iteration rather than store it at the start as I think for projectiles (which are often the things added in this loop) we probably want to run their update too.

This will, I think make beam weapons more deadly as they'll hit in the first frame and not the second, and so have not degraded their power at all (if I read it all correctly) :-)